### ISSUE TITLE
`tf.keras.layers.Convolution2DTranspose` throws error when backprop

### ISSUE THREAD

**Issue by VictoriaGriffith** [tensorflow/tensorflow#57450]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

binary

### Tensorflow Version

tf 2.9.1

### Custom Code

Yes

### OS Platform and Distribution

Linux Ubuntu 20.04

### Python version

3.7

### Current Behaviour?

```shell
When I try to do back prop on Conv2dTranspose layer, strange `TypeError` is thrown. I think this bug is related to tensorflow because the error message contains tensorflow backprop ops.
```

### Standalone code to reproduce the issue

```shell
import tensorflow as tf

filters = 2
kernel_size = [3, 3]
strides = [1, 1]
padding = "same"
output_padding = None
data_format = "channels_last"
dilation_rate = [2, 2]
activation = "linear"
use_bias = True
x = tf.random.uniform([1, 5, 6, 1], minval=0, maxval=1, dtype=tf.float32)
layer = tf.keras.layers.Convolution2DTranspose(filters, kernel_size, strides=strides, padding=padding, 
                                                                      output_padding=output_padding, data_format=data_format, dilation_rate=dilation_rate, 
                                                                      activation=activation, use_bias=use_bias)


with tf.GradientTape() as g:
    g.watch(x)
    res = layer(x)
print(res.shape) # (1, 5, 6, 2)
grad = g.jacobian(res, x) # Error
```

### Relevant log output

```shell
ERROR:tensorflow:Got error while pfor was converting op name: "gradient_tape/conv2d_transpose_11/SpaceToBatchND"
op: "SpaceToBatchND"
input: "gradient_tape/Reshape"
input: "gradient_tape/conv2d_transpose_11/SpaceToBatchND/block_shape"
input: "gradient_tape/conv2d_transpose_11/SpaceToBatchND/paddings"
attr {
  key: "T"
  value {
    type: DT_FLOAT
  }
}
attr {
  key: "Tblock_shape"
  value {
    type: DT_INT64
  }
}
attr {
  key: "Tpaddings"
  value {
    type: DT_INT32
  }
}
 with inputs (<tf.Tensor 'gradient_tape/Reshape:0' shape=(1, 5, 6, 2) dtype=float32>, <tf.Tensor 'gradient_tape/conv2d_transpose_11/SpaceToBatchND/block_shape:0' shape=(2,) dtype=int64>, <tf.Tensor 'gradient_tape/conv2d_transpose_11/SpaceToBatchND/paddings:0' shape=(2, 2) dtype=int32>)
, converted inputs [WrappedTensor(t=<tf.Tensor 'gradient_tape/Reshape/pfor/Reshape:0' shape=(60, 1, 5, 6, 2) dtype=float32>, is_stacked=True, is_sparse_stacked=False), WrappedTensor(t=<tf.Tensor 'gradient_tape/conv2d_transpose_11/SpaceToBatchND/block_shape:0' shape=(2,) dtype=int64>, is_stacked=False, is_sparse_stacked=False), WrappedTensor(t=<tf.Tensor 'gradient_tape/conv2d_transpose_11/SpaceToBatchND/paddings:0' shape=(2, 2) dtype=int32>, is_stacked=False, is_sparse_stacked=False)]
Here are the pfor conversion stack traces: Input 'y' of 'Mul' Op has type int32 that does not match type int64 of argument 'x'.
ERROR:tensorflow:name: "gradient_tape/conv2d_transpose_11/SpaceToBatchND"
op: "SpaceToBatchND"
input: "gradient_tape/Reshape"
input: "gradient_tape/conv2d_transpose_11/SpaceToBatchND/block_shape"
input: "gradient_tape/conv2d_transpose_11/SpaceToBatchND/paddings"
attr {
  key: "T"
  value {
    type: DT_FLOAT
  }
}
attr {
  key: "Tblock_shape"
  value {
    type: DT_INT64
  }
}
attr {
  key: "Tpaddings"
  value {
    type: DT_INT32
  }
}

created at:
    File "/usr/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
    File "/usr/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
    File "/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py", line 16, in <module>
    app.launch_new_instance()
    File "/usr/local/lib/python3.7/dist-packages/traitlets/config/application.py", line 846, in launch_instance
    app.start()
    File "/usr/local/lib/python3.7/dist-packages/ipykernel/kernelapp.py", line 612, in start
    self.io_loop.start()
    File "/usr/local/lib/python3.7/dist-packages/tornado/platform/asyncio.py", line 132, in start
    self.asyncio_loop.run_forever()
    File "/usr/lib/python3.7/asyncio/base_events.py", line 541, in run_forever
    self._run_once()
    File "/usr/lib/python3.7/asyncio/base_events.py", line 1786, in _run_once
    handle._run()
    File "/usr/lib/python3.7/asyncio/events.py", line 88, in _run
    self._context.run(self._callback, *self._args)
    File "/usr/local/lib/python3.7/dist-packages/tornado/ioloop.py", line 758, in _run_callback
    ret = callback()
    File "/usr/local/lib/python3.7/dist-packages/tornado/stack_context.py", line 300, in null_wrapper
    return fn(*args, **kwargs)
    File "/usr/local/lib/python3.7/dist-packages/tornado/gen.py", line 1233, in inner
    self.run()
    File "/usr/local/lib/python3.7/dist-packages/tornado/gen.py", line 1147, in run
    yielded = self.gen.send(value)
    File "/usr/local/lib/python3.7/dist-packages/ipykernel/kernelbase.py", line 365, in process_one
    yield gen.maybe_future(dispatch(*args))
    File "/usr/local/lib/python3.7/dist-packages/tornado/gen.py", line 326, in wrapper
    yielded = next(result)
    File "/usr/local/lib/python3.7/dist-packages/ipykernel/kernelbase.py", line 268, in dispatch_shell
    yield gen.maybe_future(handler(stream, idents, msg))
    File "/usr/local/lib/python3.7/dist-packages/tornado/gen.py", line 326, in wrapper
    yielded = next(result)
    File "/usr/local/lib/python3.7/dist-packages/ipykernel/kernelbase.py", line 545, in execute_request
    user_expressions, allow_stdin,
    File "/usr/local/lib/python3.7/dist-packages/tornado/gen.py", line 326, in wrapper
    yielded = next(result)
    File "/usr/local/lib/python3.7/dist-packages/ipykernel/ipkernel.py", line 306, in do_execute
    res = shell.run_cell(code, store_history=store_history, silent=silent)
    File "/usr/local/lib/python3.7/dist-packages/ipykernel/zmqshell.py", line 536, in run_cell
    return super(ZMQInteractiveShell, self).run_cell(*args, **kwargs)
    File "/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py", line 2855, in run_cell
    raw_cell, store_history, silent, shell_futures)
    File "/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py", line 2881, in _run_cell
    return runner(coro)
    File "/usr/local/lib/python3.7/dist-packages/IPython/core/async_helpers.py", line 68, in _pseudo_sync_runner
    coro.send(None)
    File "/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py", line 3058, in run_cell_async
    interactivity=interactivity, compiler=compiler, result=result)
    File "/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py", line 3249, in run_ast_nodes
    if (await self.run_code(code, result,  async_=asy)):
    File "/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py", line 3326, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
    File "<ipython-input-12-d34c7de5c6ed>", line 22, in <module>
    grad = g.jacobian(res, x) # Error
    File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/eager/backprop.py", line 1183, in jacobian
    parallel_iterations=parallel_iterations)
    File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/ops/parallel_for/control_flow_ops.py", line 202, in pfor
    outputs = f()
    File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/util/traceback_utils.py", line 150, in error_handler
    return fn(*args, **kwargs)
    File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/eager/def_function.py", line 915, in __call__
    result = self._call(*args, **kwds)
    File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/eager/def_function.py", line 963, in _call
    self._initialize(args, kwds, add_initializers_to=initializers)
    File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/eager/def_function.py", line 786, in _initialize
    *args, **kwds))
    File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/eager/function.py", line 2983, in _get_concrete_function_internal_garbage_collected
    graph_function, _ = self._maybe_define_function(args, kwargs)
    File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/eager/function.py", line 3292, in _maybe_define_function
    graph_function = self._create_graph_function(args, kwargs)
    File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/eager/function.py", line 3140, in _create_graph_function
    capture_by_value=self._capture_by_value),
    File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/func_graph.py", line 1161, in func_graph_from_py_func
    func_outputs = python_func(*func_args, **func_kwargs)
    File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/eager/def_function.py", line 677, in wrapped_fn
    out = weak_wrapped_fn().__wrapped__(*args, **kwds)
    File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/func_graph.py", line 1143, in autograph_handler
    user_requested=True,
    File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/ops/parallel_for/control_flow_ops.py", line 185, in f
    iters,
    File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/ops/parallel_for/control_flow_ops.py", line 281, in _pfor_impl
    loop_fn_outputs = loop_fn(loop_var)
    File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/eager/backprop.py", line 1173, in loop_fn
    unconnected_gradients=unconnected_gradients)
    File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/eager/backprop.py", line 1087, in gradient
    unconnected_gradients=unconnected_gradients)
    File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/eager/imperative_grad.py", line 73, in imperative_grad
    compat.as_str(unconnected_gradients.value))
    File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/eager/backprop.py", line 156, in _gradient_function
    return grad_fn(mock_op, *out_grads)
    File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/ops/array_grad.py", line 951, in _BatchToSpaceNDGrad
    array_ops.space_to_batch_nd(grad, op.inputs[1], op.inputs[2]), None, None
    File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/ops/gen_array_ops.py", line 9995, in space_to_batch_nd
    paddings=paddings, name=name)
    File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/op_def_library.py", line 742, in _apply_op_helper
    attrs=attr_protos, op_def=op_def)
    File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/func_graph.py", line 695, in _create_op_internal
    compute_device)
    File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/ops.py", line 3784, in _create_op_internal
    op_def=op_def)
    File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/ops.py", line 2175, in __init__
    self._traceback = tf_stack.extract_stack_for_node(self._c_op)

(1, 5, 6, 2)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-12-d34c7de5c6ed> in <module>
     20     res = layer(x)
     21 print(res.shape) # (1, 5, 6, 2)
---> 22 grad = g.jacobian(res, x) # Error

3 frames
/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/func_graph.py in autograph_handler(*args, **kwargs)
   1145           except Exception as e:  # pylint:disable=broad-except
   1146             if hasattr(e, "ag_error_metadata"):
-> 1147               raise e.ag_error_metadata.to_exception(e)
   1148             else:
   1149               raise

TypeError: in user code:

    File "/usr/local/lib/python3.7/dist-packages/six.py", line 703, in reraise
        raise value

    TypeError: Input 'y' of 'Mul' Op has type int32 that does not match type int64 of argument 'x'.
```
</details>

**Comment by tilakrayal** [comment]
@VictoriaGriffith,
As per the [documentation](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Conv2DTranspose?version=nightly#expandable-1) the stride and dilation_rate should be same, if strides=1,dilation_rate=(1,1) or if strides=2,dilation_rate=(2,2).
Could you please check and try to change the execute as mentioned and let us know if the issue still persists. Thank you!

**Comment by VictoriaGriffith** [comment]
Hi @tilakrayal , I don't think the document states that stride and dilation_rate should be same. Here is the related argument :

> dilation_rate: an integer, specifying the dilation rate for all spatial dimensions for dilated convolution. Specifying different dilation rates for different dimensions is not supported. **Currently, specifying any dilation_rate value != 1 is incompatible with specifying any stride value != 1**.

I think this means that only one of `stride` and `dilation_rate` can be != 1.

**Comment by VictoriaGriffith** [comment]
Plus, I checked pytorch  [torch.nn.ConvTranspose2d](https://pytorch.org/docs/stable/generated/torch.nn.ConvTranspose2d.html), and it supports strides = 1 and dilation_rate = 2:

```
import torch
from torch import nn
m = nn.ConvTranspose2d(1, 2, (3, 3), stride=(1, 1), dilation=[2,2])
input = torch.randn(1, 1, 5, 6)
output = m(input)
print(output.size())
```

**Comment by VictoriaGriffith** [comment]
Even if TensorFlow decides not to support this feature, proper error message should be thrown during the forward pass, currently `res = layer(x)` succeeds and output a `(1, 5, 6, 2)`-shape tensor, and then back prop fails.

**Comment by tilakrayal** [comment]
@sachinprasadhs,
I was able to reproduce the issue on tensorflow v2.8, v2.9 and nightly. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/1d8dc35bd961e641cd8e46849f8fab5a/untitled550.ipynb).

**Comment by oawxkw** [comment]
I also observed the following API aliases or similar APIs can cause the same issue.
Users should be cautious when using them on both CPU and GPU up to tensorflow 2.13.0 (v2.13.0-rc2-7-g1cb1a030a62).

- `tf.keras.layers.Convolution2DTranspose`, `tf.keras.layers.Conv2DTranspose`, `tf.compat.v1.keras.layers.Convolution2DTranspose`, `tf.compat.v1.keras.layers.Conv2DTranspose`
- `tf.compat.v1.layers.Conv2DTranspose`

It seems to be fixed in tensorflow 2.14.0-rc1 (v2.14.0-rc0-34-gdd01672d9a9) and nightly (2.15.0-dev20230906) versions.

**Comment by tilakrayal** [comment]
@VictoriaGriffith,
I tried to execute the above mentioned code on both [CPU](https://colab.research.google.com/gist/tilakrayal/b69fb21927a6d100f8180f1b5ae84e9f/untitled1434.ipynb) and [GPU](https://colab.research.google.com/gist/tilakrayal/42c7b2c863b2f53fa315219434dfd670/untitled1435.ipynb) with tensorflow v2.14 and it was executed without any issue/error. Kindly find the gist of it here. Thank you!
